src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorMathLibrary.java

```diff
@@ -0,0 +1,323 @@
+            Class<?> etype = vspecies.elementType();
+            if (etype != float.class && etype != double.class) {
+                return false; // only FP element types are supported
+            }
+            int maxLaneCount = VectorSupport.getMaxLaneCount(vspecies.elementType());
+            if (vspecies.length() > maxLaneCount) {
+                return false; // lacking vector support
+            }
+            if (vspecies == DoubleVector.SPECIES_64) {
+                return false; // 64-bit double vectors are not supported
+            }
+            if (op == TANH) {
+                return false; // skip due to performance considerations
+            }
+            return true;
+        }
+    }
+
+    private static final int SIZE = VectorSupport.VECTOR_OP_MATHLIB_LAST - VectorSupport.VECTOR_OP_MATHLIB_FIRST + 1;
+
+    private record Entry<T> (String name, MemorySegment entry, T impl) {}
+
+    private static final @Stable Entry<?>[][][] LIBRARY_ENTRIES = new Entry<?>[SIZE][LaneType.SK_LIMIT][VectorShape.SK_LIMIT]; // OP x SHAPE x TYPE
+
+    @ForceInline
+    private static <T> Entry<T> lookup(Operator op, int opc, VectorSpecies<?> vspecies, IntFunction<T> implSupplier) {
+        int idx = opc - VectorSupport.VECTOR_OP_MATHLIB_FIRST;
+        int elem_idx = ((AbstractSpecies<?>)vspecies).laneType.switchKey;
+        int shape_idx = vspecies.vectorShape().switchKey;
+        @SuppressWarnings({"unchecked"})
+        Entry<T> entry = (Entry<T>)LIBRARY_ENTRIES[idx][elem_idx][shape_idx];
+        if (entry == null) {
+            entry = constructEntry(op, opc, vspecies, implSupplier);
+            LIBRARY_ENTRIES[idx][elem_idx][shape_idx] = entry;
+        }
+        return entry;
+    }
+
+    @DontInline
+    private static
+    <E,T>
+    Entry<T> constructEntry(Operator op, int opc, VectorSpecies<E> vspecies, IntFunction<T> implSupplier) {
+        if (LIBRARY.isSupported(op, vspecies)) {
+            String symbol = LIBRARY.symbolName(op, vspecies);
+            try {
+                MemorySegment addr = LOOKUP.findOrThrow(symbol);
+                debug("%s %s => 0x%016x\n", op, symbol, addr.address());
+                T impl = implSupplier.apply(opc); // TODO: should call the very same native implementation eventually (once FFM API supports vectors)
+                return new Entry<>(symbol, addr, impl);
+            } catch (RuntimeException e) {
+              throw new InternalError("not supported: " + op + " " + vspecies + " " + symbol, e);
+            }
+        } else {
+            return new Entry<>(null, MemorySegment.NULL, implSupplier.apply(opc));
+        }
+    }
+
+    @ForceInline
+    /*package-private*/ static
+    <E, V extends Vector<E>>
+    V unaryMathOp(Unary op, int opc, VectorSpecies<E> vspecies,
+                  IntFunction<VectorSupport.UnaryOperation<V,?>> implSupplier,
+                  V v) {
+        var entry = lookup(op, opc, vspecies, implSupplier);
```

Comment:
Seems there is another issue for riscv here.
If the rvv extension is not supported on the running machine, it will still generate the code using rvv, this should lead to a crash at runtime?

Reply:
In previous code, we use `UseRVV` to detect if rvv extension is supported.

On the other hand, user can choose to disable UseRVV if they want even if rvv extension is supported on the running machine.

Reply:
Does the following check catch `UseRVV == false` case on RISC-V? 
```
public boolean isSupported(Operator op, VectorSpecies<?> vspecies) {
...
            int maxLaneCount = VectorSupport.getMaxLaneCount(vspecies.elementType());
            if (vspecies.length() > maxLaneCount) {
                return false; // lacking vector support
            }
...
```

Reply:
Thank you for updating! Looks good for riscv. I have ran some basic tests for vector API, passed. I did not ran benchmark, as riscv & aarch64 share the same way to bridge from java to sleef.

> Does the following check catch UseRVV == false case on RISC-V?

Yes. If you don't mind, an explicit comment might be helpful. As to me "lacking vector support" here means the vector length is not large enough, but it's quite subjective, so you are on the call.

> FTR both VectorSupport.getMaxLaneCount() and CPUFeatures don't rely on raw list of ISA extensions CPU supports, but only those reported by the JVM. So, if some feature support is disabled on JVM side, it won't be reported by VM_Version and, hence, CPUFeatures.

I'm fine with this.

Reply:
Thanks, I added some clarifications in the comments.